src/fu-engine.c

```diff
@@ -785,6 +785,106 @@ fu_engine_get_component_by_guids (FuEngine *self, FuDevice *device)
 	return NULL;
 }
 
+static XbNode *
+fu_engine_verify_from_local_metadata (FuEngine *self,
+				      FuDevice *device,
+				      XbSilo *silo,	/* has to outlive function */
+				      GError **error)
+{
+	g_autofree gchar *fn = NULL;
+	g_autofree gchar *localstatedir = NULL;
+	g_autofree gchar *xpath = NULL;
+	g_autoptr(GFile) file = NULL;
+	g_autoptr(XbBuilder) builder = xb_builder_new ();
+	g_autoptr(XbBuilderSource) source = xb_builder_source_new ();
+
+	localstatedir = fu_common_get_path (FU_PATH_KIND_LOCALSTATEDIR_PKG);
+	fn = g_strdup_printf ("%s/verify/%s.xml",
+			      localstatedir,
+			      fu_device_get_id (device));
+	file = g_file_new_for_path (fn);
+	if (!g_file_query_exists (file, NULL)) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_FOUND,
+			     "failed to find %s", fn);
+		return NULL;
+	}
+
+	if (!xb_builder_source_load_file (source, file,
+					  XB_BUILDER_SOURCE_FLAG_NONE,
+					  NULL, error))
+		return NULL;
+	xb_builder_import_source (builder, source);
+	silo = xb_builder_compile (builder,
```

Comment:
Isn't this a memory leak?

Reply:
(specifically because of the call to `xb_silo_new` earlier on with this same pointer)

Reply:
Yes, and it was before! Will fix in a smarter way, good catch.

Reply:
`g_object_set_data_full` to the rescue. We can't unconditionally ref the silo in the node else we get a dep loop and just leak tons of memory.